### DISCUSSION TITLE
conditional model association

### DISCUSSION THREAD

**Discussion by kurbhatt** [sequelize/sequelize#14630] [Help & Questions]
## Issue Creation Checklist

[x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Issue Description

Is there anyway in sequelize to have a include model association conditional?

Suppose I have list of fields to be fetched, if in that list the associated model available then only it's include should be added else not. Like this I may have multiple associated models in any model.

**ModelA:**

    @Table({  timestamps: false, freezeTableName: true,  tableName: "modela", schema: "scm" })
    export class ModelA extends Model<ModelA> {
    	@PrimaryKey
    	@Field()
    	@Column( {field: 'field1' })
    	@ForeignKey(() => ModelB)
    	field1: number;
    
    	@Field()
    	@Column( {field: 'field2' })
    	field2: string;
    
    	@Field({nullable: true})
        @BelongsTo(() => ModelB, { foreignKey: 'field1', targetKey:'fieldB1'})
        modelB: ModelB
    }

**ModelB:**

    @Table({  timestamps: false, freezeTableName: true,  tableName: "modelb", schema: "scm" })
    export class ModelB extends Model<ModelB> {
    	@PrimaryKey
    	@Field()
    	@Column( {field: 'field1' })
    	fieldB1: number;
    
    	@Field()
    	@Column( {field: 'field2' })
    	fieldB2: string;
    }

Now I am fetching specific fields through sequelize query using attributes.
I have list of fields like: **field1, modelB.fieldB1**

now in this case the query will be something like:

    return this.repo.findOne({
    	attributes: fields,
    	include: [
    	  { 
    	    model: ModelB
    	  }
    	],
    	where: { field1 },
    });

Here I am adding include as there's field required in select query.
But suppose I have field list like: **field1, field2**

in this case query should be something like:

    return this.repo.findOne({
    	attributes: fields,
    	where: { field1 },
    });

This I can do conditionally by having a if block which have 2 separate code sections, 1 with include another without include.
So Is there any library specified configuration available that will conditionally enable/disable include based on some flag or condition?

### Additional context

Add any other context or screenshots about the issue here.

### Is this issue dialect-specific?

- [ ] No. This issue is relevant to Sequelize as a whole.
- [ ] Yes. This issue only applies to the following dialect(s): XXX, YYY, ZZZ
- [x] I don't know.

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [x] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by ephys** [comment]
You can use the [`beforeFind`](https://sequelize.org/docs/v6/other-topics/hooks/) hook to modify your `findAll` query and automatically add the include

**Reply by kurbhatt** [reply]
There's no **beforeFind** hook in given article. Do I need to create custom? If there's any sample or some article then please let me know.
Over here we can do same thing which we can do before executing query by dynamically creating **attributes** and **includes**, right?
I have created a method which will dynamically generates **attributes** and **include** based on the requested fields and passing them while querying. Same thing we can do in hook as well, nothing more than that. Right?

**Reply by ephys** [reply]
The documentation on hooks isn't brilliant but the hook is declared here: https://github.com/sequelize/sequelize/blob/v6/src/hooks.js#L31, as referenced here: https://sequelize.org/docs/v6/other-topics/hooks/#available-hooks

You can declare it like this: https://sequelize.org/docs/v6/other-topics/hooks/#declaring-hooks

Then you'll need to experiment by checking what the hook received, you can modify the query that way.

```js
class User extends Model {}
User.init({
  // ...
}, {
  hooks: {
    beforeFindAll: (options) => {
      // modify the query
      options.include = [
        // ...
      ];
    },
  },
  sequelize
});
```

**Reply by kurbhatt** [reply]
Got it. I want the hook to be executed on some queries only so before that query I can add hook and after executing query I can remove hook, right?
```
Model1.addHook('beforeFind', (options) => {
  console.log(" BEFORE HOOOOOOOOK")
  console.log(options.include);
  let incds = options.include;
  if (incds['length'] > 0) {
    incds[0].attributes = ['uId']; // whatever was there, just fetch this field only
  }
  console.log(" AFTER HOOOOOOOOK")
  console.log(options.include);
});

// query execution

Model1.removeHook('beforeFind');
```

Is this correct approach?
